include/stl2/detail/memory/uninitialized_copy.hpp

```diff
@@ -41,24 +40,38 @@ STL2_OPEN_NAMESPACE {
 		return {std::move(first), std::move(result)};
 	}
 
+	///////////////////////////////////////////////////////////////////////////
+	// uninitialized_copy [Extension]
+	//
 	template <InputRange Rng, __NoThrowForwardIterator O>
 	requires
-		Constructible<value_type_t<O>, reference_t<iterator_t<Rng>>> &&
-		__ReferenceTo<O, value_type_t<O>>
+		Constructible<value_type_t<O>, reference_t<iterator_t<Rng>>>
 	tagged_pair<tag::in(safe_iterator_t<Rng>), tag::out(O)>
 	uninitialized_copy(Rng&& rng, O result)
 	{
 		return __stl2::uninitialized_copy(
 			__stl2::begin(rng), __stl2::end(rng), result);
 	}
 
+	///////////////////////////////////////////////////////////////////////////
+	// uninitialized_copy [Extension]
+	//
+	template <InputRange IRng, __NoThrowForwardRange ORng>
+	requires
+		Constructible<value_type_t<iterator_t<ORng>>, reference_t<iterator_t<IRng>>>
+	tagged_pair<tag::in(safe_iterator_t<IRng>), tag::out(iterator_t<ORng>)>
+	uninitialized_copy(IRng&& irng, ORng& orng)
+	{
+		return __stl2::uninitialized_copy(
+			__stl2::begin(irng), __stl2::end(irng), __stl2::begin(orng));
```

Comment:
We should decide what to do for a range size mismatch. Being able to handle that safely is a major advantage of taking two ranges rather than range-and-a-half.

Reply:
How likely are contracts to hit C++20? I'm in favour of `[[expects: ranges::size(irng) <= ranges::size(orng)]]` over throwing for these algorithms in particular.

Reply:
I agree that precondition is best here, despite that it can't always be checked because `irng` is single-pass. (You should use `distance` instead of `size` in the precondition.) Throwing is for failure to meet postconditions due to unforeseen (exceptional) circumstances, not for communicating "You gave me garbage input" to the caller.

Reply:
I'm not particularly convinced that a smaller output range necessarily constitutes garbage input. "Copy until either range is exhausted" seems to be a plausible behavior to me.

Reply:
Added bonus: this solves the signed vs unsigned problem.

Reply:
> "Copy until either range is exhausted" seems to be a plausible behavior to me.

I can see arguments for any of:
1. "Copy until either range is exhausted" (no precondition on range lengths)
2. "Always copy every element of the input" (|input| <= |output|)
3. "Always write to every element of the output" (|input| >= |output|)
4. "Always copy each element of the input into each element of the output" (|input| == |output|)

(2) most closely matches the behavior of output iterator algorithms - they simply assume that the output range has enough space to copy every input element. We need a clear benefit to disregard consistency of `copy(irange, orange)` with `copy(first, last, result)`.

I think the "clear benefit" of choosing (1) is that a caller can more easily adapt an algorithm to each of the other semantics than if we were to choose (2), (3), or (4). Implementing (1) atop (4), for example, is just barely feasible when `irange` and `orange` are both single-pass - certainly not a job for someone who isn't an expert in range adaptors. Implementing any of the others atop (1), however, is as simple as comparing the iterators returned from the algorithm with the end sentinels of the ranges and `assert`ing.

I'm a bit nervous making this level of policy decision in a code review without getting wider community input - realistically we're talking about more than just `uninitialized_copy`. @ericniebler, thoughts?

Reply:
I made an issue here: https://github.com/ericniebler/stl2/issues/556